### PR TITLE
multi-gpu ppo

### PR DESCRIPTION
**Pull request by dongruoping**
add multi-gpu ppo training

### CODE REVIEW

File: ml-agents/mlagents/trainers/ppo/models.py [hunk 1]
```diff
@@ -1,12 +1,115 @@
...
+# Variable scope in which created variables will be placed under
+TOWER_SCOPE_NAME = "tower"
+
 logger = logging.getLogger("mlagents.trainers")
 
 
+class PPOMultiGPUModel(object):
+    def __init__(
+        self,
+        brain,
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/models.py]: Seems like you should be able to use `*args` and `**kwargs` and pass these through to prevent duplication, right?  The only argument specific to this model is `devices`.  So then you can generically pass in the args so we don't have to change both places when we change PPO.

File: ml-agents/mlagents/trainers/ppo/models.py [hunk 2]
```diff
@@ -1,12 +1,115 @@
...
 import tensorflow as tf
 from mlagents.trainers.models import LearningModel
 
+# Variable scope in which created variables will be placed under
+TOWER_SCOPE_NAME = "tower"
+
 logger = logging.getLogger("mlagents.trainers")
 
 
+class PPOMultiGPUModel(object):
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/models.py]: I'd recommend moving this class into its own file.

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 3]
```diff
@@ -461,11 +473,13 @@ def update_policy(self):
...
                 len(self.training_buffer.update_buffer["actions"]) // n_sequences
             ):
-                start = l * n_sequences
-                end = (l + 1) * n_sequences
-                run_out = self.policy.update(
-                    buffer.make_mini_batch(start, end), n_sequences
-                )
+                mini_batches = []
+                n_sequences_per_device = n_sequences // len(self.devices)
+                for d in range(len(self.devices)):
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/trainer.py]: Rather than `d` and `l` maybe these could be `seq_idx` and `device_idx`?

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 4]
```diff
@@ -461,11 +473,13 @@ def update_policy(self):
             for l in range(
                 len(self.training_buffer.update_buffer["actions"]) // n_sequences
             ):
-                start = l * n_sequences
-                end = (l + 1) * n_sequences
-                run_out = self.policy.update(
-                    buffer.make_mini_batch(start, end), n_sequences
-                )
+                mini_batches = []
+                n_sequences_per_device = n_sequences // len(self.devices)
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/trainer.py]: Is there a way we could keep the `Trainer` class from knowing about devices?  Seems like we're still passing the same number of sequences total, so we could probably do this split at the `Policy` level.

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 5]
```diff
@@ -514,3 +528,13 @@ def get_gae(rewards, value_estimates, value_next=0.0, gamma=0.99, lambd=0.95):
     delta_t = rewards + gamma * value_estimates[1:] - value_estimates[:-1]
     advantage = discount_rewards(r=delta_t, gamma=gamma * lambd)
     return advantage
+
+
+def get_devices(multi_gpu):
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/trainer.py]: This seems like a utility function.  I'd probably put it in some "tensorflow util" module or put it with the multi gpu model file.

File: ml-agents/mlagents/trainers/ppo/models.py (hunk 2 above)
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/models.py]: This doesn't need to subclass from `object`, It can just be `class PPOMultiGPUModel:`

File: ml-agents/mlagents/trainers/ppo/trainer.py (hunk 5 above)
Comment (dongruoping) [reply ml-agents/mlagents/trainers/ppo/trainer.py]: created a new multi gpu model file and moved this there

File: ml-agents/mlagents/trainers/ppo/multi_gpu_model.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            avg_grad = tf.reduce_mean(tf.stack(grads), 0)
+            var = grad_and_vars[0][1]
+            average_grads.append((avg_grad, var))
+        return average_grads
+
+
+def get_devices(multi_gpu):
+    local_device_protos = device_lib.list_local_devices()
+    devices = [x.name for x in local_device_protos if x.device_type == "GPU"]
+    if len(devices) < 1:
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/multi_gpu_model.py]: Why is this case different than the `not multi_gpu` case?

In any case, I'd change this to be `get_gpu_devices` and return an empty list if none.

File: ml-agents/mlagents/trainers/ppo/policy.py [hunk 7]
```diff
@@ -29,24 +30,42 @@ def __init__(self, seed, brain, trainer_params, is_training, load):
 
         reward_signal_configs = trainer_params["reward_signals"]
 
+        self.devices = get_devices(multi_gpu)
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/policy.py]: Rather than passing `multi_gpu` here if you list all the GPUs you can just say `if len(self.gpus) > 1 and multi_gpu: [...]` below.

File: ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+        device_batch_size = n_sequences // len(self.devices)
+        device_batches = []
+        for i in range(len(self.devices)):
+            device_batches.append(
+                {k: v[i : i + device_batch_size] for k, v in mini_batch.items()}
+            )
+
+        assert len(device_batches) == len(towers)
+        for batch, tower in zip(device_batches, towers):
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py]: It seems that the logic below for constructing the feed dict is common between `PPOPolicy` and `MultiGPUPPOPolicy`.  I'd recommend you factor the logic into a function like `PPOPolicy.construct_feed_dict(model, minibatch, num_sequences)` and then use that here to gather the per-graph components of the feed dict so we don't duplicate logic.

File: ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+                                m_size=self.m_size,
+                                seed=seed,
+                                stream_names=list(reward_signal_configs.keys()),
+                                vis_encode_type=trainer_params["vis_encode_type"],
+                            )
+                        )
+                        self.towers[-1].create_ppo_optimizer()
+
+            # Create reward signals
+            for reward_signal, config in reward_signal_configs.items():
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py]: This reward signal logic and the pretrainer logic below is shared with the base class `__init__`.  It would probably be best to extract it somehow so we don't have to duplicate it.

File: ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+                    default_batch_size=trainer_params["batch_size"],
+                    default_num_epoch=trainer_params["num_epoch"],
+                    **trainer_params["pretraining"],
+                )
+            else:
+                self.bc_module = None
+
+        self.model = self.towers[0]
+
+        if load:
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/multi_gpu_policy.py]: Same as above, we duplicate a lot of logic here.

File: ml-agents/mlagents/trainers/tests/test_bcmodule.py [hunk 11]
```diff
@@ -77,7 +77,7 @@ def create_ppo_policy_with_bc_mock(
     trainer_parameters["pretraining"]["demo_path"] = (
         os.path.dirname(os.path.abspath(__file__)) + "/" + demo_file
     )
-    policy = PPOPolicy(0, mock_brain, trainer_parameters, False, False)
+    policy = PPOPolicy(0, mock_brain, trainer_parameters, False, False, False)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/tests/test_bcmodule.py]: Did you run the Pytest ? You added a False here but I don't know why, it seems to break the tests.

File: ml-agents/mlagents/trainers/tf_policy.py [hunk 12]
```diff
@@ -60,6 +60,7 @@ def __init__(self, seed, brain, trainer_parameters):
         self.graph = tf.Graph()
         config = tf.ConfigProto()
         config.gpu_options.allow_growth = True
+        config.allow_soft_placement = True
```
[the remark this answers is not in the document]
Comment (dongruoping) [reply ml-agents/mlagents/trainers/tf_policy.py]: Added now

File: ml-agents/mlagents/trainers/tests/test_bcmodule.py (hunk 11 above)
Comment (dongruoping) [reply ml-agents/mlagents/trainers/tests/test_bcmodule.py]: Was still in progress of a major change. Fixed now.
